### PR TITLE
feat: SDK-driven MCP tool registrar (INT-351)

### PR DESCRIPTION
**Pull request by darvell-thenvoi**
## Summary

Phase 3 of INT-338. Replaces handwritten per-tool `@mcp.tool()` registrations
with a scope-filtered loop over the SDK's `iter_tool_definitions(...)`. Each
`ToolDefinition` becomes a FastMCP tool whose handler dispatches to the Phase 1
`HumanTools` / `AgentTools` from `thenvoi-sdk-python`.

- New: `src/thenvoi_mcp/tools/registrar.py` — `register_tools(mcp, config)`.
- New: `tests/unit/test_registrar.py` — 28 tests.
- New: `tests/integration/test_forwarding.py` — 11 in-process MCP round-trip tests.
- Modified: `src/thenvoi_mcp/server.py` — calls `register_tools(mcp, config)` after the existing `load_tools()` invocation. Old handwritten handlers continue to register in parallel (SDK tool names are `thenvoi_`-prefixed and do not collide). Phase 4 (INT-352) deletes the handwritten handlers.

## Spec deviation (resolved with reviewer)

The Phase 3 spec text told the registrar to classify agent tools by looking
for a `room_id` field on `ToolDefinition.input_model.model_fields`. That
classifier does not work for agent tools: `AgentTools` is room-scoped via
its constructor, so the SDK input models only cover method arguments, not
the construction-time room id. Putting `room_id` on the input model would
create a mismatch between the input schema and the `AgentTools` method
signature.

**Resolution**: the registrar adds `chat_id` to the advertised schema of
every agent room-bound tool, with `AliasChoices(\"chat_id\", \"room_id\")` so
both names work at the wire. This keeps the SDK clean (pure tool logic),
matches today's MCP wire format exactly (zero breaking change for existing
MCP consumers), and still satisfies the spec's intent (scope-filtered
registry + pinned-room field stripping).

The set of agent room-bound tools is enumerated in
`AGENT_ROOM_BOUND_TOOL_NAMES`, derived from today's agent handlers that
currently take `chat_id` in `src/thenvoi_mcp/tools/agent/*.py`. Rationale
is documented in the module docstring.

The human-side classifier is unchanged: human input models already carry a
`chat_id` field where applicable (Phase 1 derived them from `HumanTools`
method signatures), so `model_fields`-based detection works on that surface.

## Old handler coexistence

Chose option (a): keep the old handwritten handler imports and add the
new registrar after them. FastMCP logs a warning and keeps the
first-registered tool on duplicate names; since SDK tool names are
`thenvoi_`-prefixed (e.g. `thenvoi_list_my_contacts`) and today's
handwritten handler names are unprefixed (e.g. `list_my_contacts`), there
is no collision. Both surfaces coexist during the transition. Phase 4
(INT-352) deletes the handwritten handlers wholesale.

## Dependency

Depends on thenvoi/thenvoi-sdk-python#264 (Phase 1, INT-349). Local
development pin is in the PR checkout notes; the committed
`pyproject.toml` / `uv.lock` are unchanged from Phase 2.

## Test plan

- [x] `uv run pytest tests/unit/test_registrar.py --no-cov -v` — 28 passed
- [x] `uv run pytest tests/integration/test_forwarding.py --no-cov -v` — 11 passed
- [x] `uv run pytest tests/unit/ tests/integration/test_forwarding.py --no-cov` — 86 passed
- [x] `uv run pre-commit run --all-files` — ruff, ruff-format, pyrefly all green
- [ ] Smoke-test against live API after Phase 1 (INT-349) merges

## Linked

- Parent: INT-338
- Phase 1 (SDK): thenvoi/thenvoi-sdk-python#264 (INT-349)
- Phase 2 (this repo): #99 (INT-350, stacked base)
- Blocks: INT-352 (Phase 4, handwritten handler deletion)

### CODE REVIEW

File: src/thenvoi_mcp/tools/registrar.py [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+    kwargs: dict[str, Any],
+) -> str:
+    """The actual async dispatch body shared by every generated handler."""
+    reset_agent_tools_cache(ctx)
+
+    # Inject pinned room id BEFORE validation so the input model's chat_id
+    # field is populated from the pin even though it is hidden from the
+    # advertised schema.
+    if pinned_room_id is not None and (is_agent_room_bound or is_human_room_bound):
+        kwargs.setdefault("chat_id", pinned_room_id)
```
Comment (amit-gazal-band) [inline src/thenvoi_mcp/tools/registrar.py]: **[Required]** Pin is bypassable by caller.

`setdefault` only fills `chat_id` when the key is absent. A caller that sends `chat_id` (or its `room_id` alias via AliasChoices at the generated arg-model layer) overrides the pin. If an operator runs `--room-id X` to lock the MCP to one room, pin enforcement must be authoritative.

Suggest:
```python
kwargs["chat_id"] = pinned_room_id  # pin wins, always
```

Also add a test: "pinned mode + caller sends chat_id=Y → dispatch uses pinned room, not Y".

File: src/thenvoi_mcp/tools/registrar.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+    try:
+        validated = input_model.model_validate(kwargs)
+    except Exception as exc:
```
Comment (amit-gazal-band) [inline src/thenvoi_mcp/tools/registrar.py]: **[Required]** Bare `except Exception` violates `.claude/rules/01-error-handling.md`.

Project rule mandates catching `pydantic.ValidationError` specifically and formatting per-field. Bare `Exception` here swallows unrelated bugs (custom validator AttributeError, etc.) as "invalid arguments".

```python
from pydantic import ValidationError

try:
    validated = input_model.model_validate(kwargs)
except ValidationError as exc:
    errors = "; ".join(f"{err['loc'][0]}: {err['msg']}" for err in exc.errors())
    raise ValueError(f"Invalid arguments for {tool_name}: {errors}") from exc
```

File: src/thenvoi_mcp/tools/registrar.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+            # SDK's ``AgentTools`` is constructor-scoped, but such tools
+            # only touch ``self.rest`` — we can safely construct with the
+            # pinned room id (or an empty placeholder if none is set) and
+            # dispatch.
+            tools_instance = get_agent_tools(ctx, pinned_room_id or "")
+    else:
+        tools_instance = get_human_tools(ctx)
+
+    if tools_instance is None:
+        raise RuntimeError(
```
Comment (amit-gazal-band) [inline src/thenvoi_mcp/tools/registrar.py]: **[Required]** `RuntimeError` for a config-side condition — prefer `ValueError` per project convention (`.claude/rules/01-error-handling.md`: missing/unavailable config → `ValueError`, not `RuntimeError`).

"SDK not installed or no credential configured for this scope" is a configuration failure, user-facing. `ValueError` here. The second `raise RuntimeError` at line 314 (`method not found`) is an internal invariant violation — `RuntimeError` acceptable there, or keep `AttributeError`.

File: src/thenvoi_mcp/tools/registrar.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+                    f"{tool_name}: missing chat_id (or room_id) for room-bound tool"
+                )
+            tools_instance = get_agent_tools(ctx, chat_id)
+        else:
+            # Room-less agent tool (e.g. ``thenvoi_create_chatroom``). The
+            # SDK's ``AgentTools`` is constructor-scoped, but such tools
+            # only touch ``self.rest`` — we can safely construct with the
+            # pinned room id (or an empty placeholder if none is set) and
+            # dispatch.
+            tools_instance = get_agent_tools(ctx, pinned_room_id or "")
```
Comment (amit-gazal-band) [inline src/thenvoi_mcp/tools/registrar.py]: **[Required]** `pinned_room_id or ""` passes empty string to `AgentTools` for room-less agent tools (e.g. `thenvoi_create_chatroom`).

Works today because room-less methods don't touch `self.room_id`, but invariant is fragile and not covered by any test. Options:
1. Make `get_agent_tools`/`AgentTools` accept `None` and pass `pinned_room_id` directly.
2. Keep empty-string fallback but add test exercising a room-less agent tool dispatch end-to-end, and a comment on `AgentTools` that empty `room_id` is legal iff only room-less methods are called.

Prefer (1).

**Review by amit-gazal-thenvoi — CHANGES_REQUESTED** [review]
Required changes — see inline. Issue #1 is a pin-bypass (correctness + intent), #2 violates project error-handling rule, #3 inconsistent with project convention, #4 smelly invariant worth nailing down.

**Comment by darvell-thenvoi** [comment]
Addressed the required registrar review items:

- pinned room id now wins over caller-provided `chat_id`
- validation now catches `ValidationError` specifically and formats per-field errors
- config/SDK availability now raises `ValueError`
- room-less agent tools pass `None` through to `AgentTools` instead of an empty string placeholder

Verified with:
- `PYTHONPATH=src uv run pytest tests/unit/test_shared.py --no-cov`
- `PYTHONPATH=src uv run pytest tests/unit/test_registrar.py tests/unit/test_shared.py --no-cov`
- `uv run ruff check src/thenvoi_mcp/tools/registrar.py src/thenvoi_mcp/shared.py tests/unit/test_registrar.py tests/unit/test_shared.py`

**Review by amit-gazal-thenvoi — APPROVED** [review]
